extensions/AWS/S3/S3Storage.cs

```diff
@@ -0,0 +1,346 @@
+    {
+        var directoryName = JoinPaths(index, documentId);
+        if (string.IsNullOrWhiteSpace(index) || string.IsNullOrWhiteSpace(documentId) || string.IsNullOrWhiteSpace(directoryName))
+        {
+            throw new DocumentStorageException("The index, or document ID, or directory name is empty, stopping the process to prevent data loss");
+        }
+
+        await this.DeleteObjectsByPrefixAsync(directoryName, cancellationToken).ConfigureAwait(false);
+    }
+
+    /// <inheritdoc />
+    public async Task DeleteDocumentDirectoryAsync(
+        string index,
+        string documentId,
+        CancellationToken cancellationToken = default)
+    {
+        var directoryName = JoinPaths(index, documentId);
+        if (string.IsNullOrWhiteSpace(index) || string.IsNullOrWhiteSpace(documentId) || string.IsNullOrWhiteSpace(directoryName))
+        {
+            throw new DocumentStorageException("The index, or document ID, or directory name is empty, stopping the process to prevent data loss");
+        }
+
+        await this.DeleteObjectsByPrefixAsync(directoryName, cancellationToken).ConfigureAwait(false);
+    }
+
+    /// <inheritdoc />
+    public async Task WriteFileAsync(
+        string index,
+        string documentId,
+        string fileName,
+        Stream streamContent,
+        CancellationToken cancellationToken = default)
+    {
+        var directoryName = JoinPaths(index, documentId);
+        var objName = $"{directoryName}/{fileName}";
+        var len = streamContent.Length;
+
+        this._log.LogTrace("Writing object {0} ...", objName);
+
+        if (streamContent.Length == 0)
+        {
+            this._log.LogWarning("The file {0} is empty", objName);
+        }
+
+        await this._client.PutObjectAsync(new PutObjectRequest
+        {
+            BucketName = this._bucketName,
+            Key = objName,
+            InputStream = streamContent
+        }, cancellationToken: cancellationToken).ConfigureAwait(false);
+
+        this._log.LogTrace("Object {0} ready, size {1}", objName, len);
+    }
+
+    /// <inheritdoc />
+    public async Task<StreamableFileContent> ReadFileAsync(
+        string index,
+        string documentId,
+        string fileName,
+        bool logErrIfNotFound = true,
+        CancellationToken cancellationToken = default)
+    {
+        var directoryName = JoinPaths(index, documentId);
+        var objName = $"{directoryName}/{fileName}";
+
+        try
+        {
+            using (var response = await this._client.GetObjectAsync(
+                new GetObjectRequest
+                {
+                    BucketName = this._bucketName,
+                    Key = objName,
+                },
+                cancellationToken
+            ).ConfigureAwait(false))
+            {
+                var memoryStream = new MemoryStream();
+                var responseStream = response.ResponseStream;
+
+                await using (responseStream.ConfigureAwait(false))
```

Comment:
@dluc is this the correct way to use `await using`? And should we also do something similar with `memoryStream`? (I am not sure when to dispose it, because it is used in the callback function of StreamableFileContent, line 197)

Reply:
not sure if that builds, combining `await using` and `ConfigureAwait(false)` in the same line is not possible as far as I know. `using` takes care of disposing the object for you, though in some cases dispose is an async method so you need to use `await using`. In your code I saw only one instance of this. If you check the postgresdbclient you might notice I split the code in two lines. In any case, I can fix this before merging the PR (right now I don't remember the exact syntax, sorry)

Reply:
I was able to build it and it actually works, it save the file to S3. I found this syntax in the codebase 
<img width="418" alt="image" src="https://github.com/microsoft/kernel-memory/assets/428201/49539cfb-52dc-4085-9019-604f10bb8cc7">

Reply:
LGTM